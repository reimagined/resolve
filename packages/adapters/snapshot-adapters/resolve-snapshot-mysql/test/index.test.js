import MySQL from 'mysql2/promise'
import createSnapshotAdapter from '../src'

describe('resolve-snapshot-mysql', () => {
  const bucketSize = 5
  let snapshotAdapter = null

  beforeEach(async () => {
    snapshotAdapter = createSnapshotAdapter({
      host: 'localhost',
      port: 3306,
      user: 'user',
      password: 'password',
      database: 'database',
      bucketSize
    })

    MySQL.createConnection().execute.mockImplementation(() => {})
  })

  afterEach(async () => {
    MySQL.createConnection().execute.mockReset()

    if (snapshotAdapter != null) {
      await snapshotAdapter.dispose()
    }
  })

  test(`"saveSnapshot" should save the snapshot every 5 times`, async () => {
    await snapshotAdapter.init()

    for (let index = 0; index < bucketSize; index++) {
      await snapshotAdapter.saveSnapshot('key', `value = ${index}`)
    }
    await snapshotAdapter.saveSnapshot('key', `value = ${bucketSize}`)

    expect(MySQL.createConnection().execute.mock.calls).toMatchSnapshot()
  })

  test(`"loadSnapshot" should load the snapshot`, async () => {
    MySQL.createConnection().execute.mockReturnValueOnce([
      [{ SnapshotContent: '"value"' }]
    ])

    const value = await snapshotAdapter.loadSnapshot('key')

    expect(MySQL.createConnection().execute.mock.calls).toMatchSnapshot()
    expect(value).toMatchSnapshot()
  })

  test(`"drop" should drop the snapshotAdapter`, async () => {
    await snapshotAdapter.dropSnapshot('key')
    expect(MySQL.createConnection().execute.mock.calls).toMatchSnapshot()
  })

  test(`"dispose" should dispose the snapshotAdapter`, async () => {
    await snapshotAdapter.init()
    await snapshotAdapter.dispose()

    try {
      await snapshotAdapter.dispose()
      snapshotAdapter = null
      return Promise.reject(new Error('Test failed'))
    } catch (error) {
      expect(error).toBeInstanceOf(Error)
      expect(error.message).toEqual('Adapter is disposed')
      snapshotAdapter = null
    }
  })
})
