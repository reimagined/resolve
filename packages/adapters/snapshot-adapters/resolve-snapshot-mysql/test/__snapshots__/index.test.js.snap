// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve-snapshot-mysql "drop" should drop the snapshotAdapter 1`] = `
Array [
  Array [
    "DELETE FROM \`__ResolveSnapshots__\`
    WHERE \`SnapshotKey\`
    LIKE 'key%'",
    Array [
      "key",
    ],
  ],
]
`;

exports[`resolve-snapshot-mysql "loadSnapshot" should load the snapshot 1`] = `
Array [
  Array [
    "SELECT \`SnapshotContent\` FROM \`__ResolveSnapshots__\`
   WHERE \`SnapshotKey\`= 'key' ",
  ],
]
`;

exports[`resolve-snapshot-mysql "loadSnapshot" should load the snapshot 2`] = `"\\"value\\""`;

exports[`resolve-snapshot-mysql "saveSnapshot" should save the snapshot every 5 times 1`] = `
Array [
  Array [
    "CREATE TABLE IF NOT EXISTS \`__ResolveSnapshots__\` (
      \`SnapshotKey\` MEDIUMBLOB NOT NULL,
      \`SnapshotContent\` LONGBLOB,
      PRIMARY KEY(\`SnapshotKey\`(255))
    )",
  ],
  Array [
    "INSERT INTO \`__ResolveSnapshots__\`(\`SnapshotKey\`, \`SnapshotContent\`)
    VALUES('key', 'value = 5')
    ON DUPLICATE KEY UPDATE
    \`SnapshotContent\` = 'value = 5'",
  ],
]
`;
